vmm/src/lib.rs

```diff
@@ -999,141 +999,161 @@ impl Vmm {
-                Response::read_from(&mut socket).ok();
-                return Err(MigratableError::MigrateSend(anyhow!(
-                    "Error during state migration"
-                )));
-            }
+                // Now pause VM
+                vm.pause()?;
+                paused = true;
+
+                // Send last batch of dirty pages
+                Self::vm_maybe_send_dirty_pages(vm, &mut socket)?;
+
+                // Capture snapshot and send it
+                let vm_snapshot = vm.snapshot()?;
+                let snapshot_data = serde_json::to_vec(&vm_snapshot).unwrap();
+                Request::state(snapshot_data.len() as u64).write_to(&mut socket)?;
+                socket
+                    .write_all(&snapshot_data)
+                    .map_err(MigratableError::MigrateSocket)?;
+                let res = Response::read_from(&mut socket)?;
+                if res.status() != Status::Ok {
+                    warn!("Error during state migration");
+                    Request::abandon().write_to(&mut socket)?;
+                    Response::read_from(&mut socket).ok();
+                    return Err(MigratableError::MigrateSend(anyhow!(
+                        "Error during state migration"
+                    )));
+                }
+
+                // Complete the migration
+                Request::complete().write_to(&mut socket)?;
+                let res = Response::read_from(&mut socket)?;
+                if res.status() != Status::Ok {
+                    warn!("Error completing migration");
+                    Request::abandon().write_to(&mut socket)?;
+                    Response::read_from(&mut socket).ok();
+                    return Err(MigratableError::MigrateSend(anyhow!(
+                        "Error completing migration"
+                    )));
+                }
+                info!("Migration complete");
+                Ok(())
+            } {
+                // Stop logging dirty pages and keep the source VM paused unpon successful migration
+                Ok(()) => {
+                    // Stop logging dirty pages
+                    vm.stop_memory_dirty_log()?;
 
-            // Complete the migration
-            Request::complete().write_to(&mut socket)?;
-            let res = Response::read_from(&mut socket)?;
-            if res.status() != Status::Ok {
-                warn!("Error completing migration");
-                Request::abandon().write_to(&mut socket)?;
-                Response::read_from(&mut socket).ok();
-                return Err(MigratableError::MigrateSend(anyhow!(
-                    "Error completing migration"
-                )));
+                    Ok(())
+                }
+                // Ensure the source VM continue to run upon unsuccessful migration
+                Err(e) => {
+                    // Stop logging dirty pages
+                    vm.stop_memory_dirty_log()?;
+
+                    if paused {
```

Comment:
Is there a function on `vm` that you can use to look at the state to avoid this bool.

Reply:
Right. I think `vm.get_state()` should do the work. PR updated.